src/main/java/io/github/fusionflux/portalcubed/data/models/PedestalButtonBlockStates.java

```diff
@@ -0,0 +1,127 @@
+
+import io.github.fusionflux.portalcubed.content.PortalCubedBlocks;
+import io.github.fusionflux.portalcubed.content.button.pedestal.PedestalButtonBlock;
+import net.fabricmc.fabric.api.datagen.v1.FabricDataOutput;
+import net.fabricmc.fabric.api.datagen.v1.provider.FabricModelProvider;
+import net.minecraft.core.Direction;
+import net.minecraft.data.models.BlockModelGenerators;
+import net.minecraft.data.models.ItemModelGenerators;
+import net.minecraft.data.models.blockstates.Condition;
+import net.minecraft.data.models.blockstates.MultiPartGenerator;
+import net.minecraft.data.models.blockstates.Variant;
+import net.minecraft.data.models.blockstates.VariantProperties;
+import net.minecraft.data.models.blockstates.VariantProperty;
+import net.minecraft.data.models.model.ModelLocationUtils;
+import net.minecraft.world.phys.Vec3;
+
+public class PedestalButtonBlockStates extends FabricModelProvider {
+	public static final VariantProperty<Float> SHIFT_X = new VariantProperty<>("portalcubed:shift_x", JsonPrimitive::new);
+	public static final VariantProperty<Float> SHIFT_Y = new VariantProperty<>("portalcubed:shift_y", JsonPrimitive::new);
+	public static final VariantProperty<Float> SHIFT_Z = new VariantProperty<>("portalcubed:shift_z", JsonPrimitive::new);
+	public static final VariantProperty<Integer> Z_ROT = new VariantProperty<>("portalcubed:rot_z", JsonPrimitive::new);
+	public static final VariantProperty<Boolean> LOCAL_Z_ROT = new VariantProperty<>("portalcubed:local_rot_z", JsonPrimitive::new);
+
+	public PedestalButtonBlockStates(FabricDataOutput output) {
+		super(output);
+	}
+
+	@Override
+	public void generateBlockStateModels(BlockModelGenerators blockStateModelGenerator) {
+		createPedestalButton(blockStateModelGenerator, PortalCubedBlocks.PEDESTAL_BUTTON);
+		createPedestalButton(blockStateModelGenerator, PortalCubedBlocks.OLD_AP_PEDESTAL_BUTTON);
+	}
+
+	public void createPedestalButton(BlockModelGenerators blockStateModelGenerator, PedestalButtonBlock pedestalButtonBlock) {
+		var regularModelId = ModelLocationUtils.getModelLocation(pedestalButtonBlock);
+		var activeModelId = ModelLocationUtils.getModelLocation(pedestalButtonBlock, "_active");
+		var centerBaseModelId = ModelLocationUtils.getModelLocation(pedestalButtonBlock, "_base_center");
+		var edgeBaseModelId = ModelLocationUtils.getModelLocation(pedestalButtonBlock, "_base_edge");
+
+		var generator = MultiPartGenerator.multiPart(pedestalButtonBlock);
+		for (var state : pedestalButtonBlock.getStateDefinition().getPossibleStates()) {
+			var face = state.getValue(PedestalButtonBlock.FACE);
+			var facing = state.getValue(PedestalButtonBlock.FACING);
+			var offset = state.getValue(PedestalButtonBlock.OFFSET);
+			boolean base = state.getValue(PedestalButtonBlock.BASE);
+
+			var variant = Variant.variant();
+			{
+				var faceAxis = face.getAxis();
+				boolean flip = face.getAxisDirection() == Direction.AxisDirection.NEGATIVE;
+				variant = switch (faceAxis) {
```

Comment:
My eyes

Reply:
can you explain

Reply:
this code is hard to read. at least make the lines not comically long
I would do it like this
```java
flip ? variant.with(...)
          .with(...)
          .with(...)
    : variant.with(...)
```
or make some methods to shorten everything